[WIP] Updates the form and configuration

agentrickard opened the pull request.

Commit:
Issue #3038421 by agentrickard, mweixel: Do not require leading slash in config

agentrickard committed:
Initial setup fieldset.

agentrickard committed:
Sets the options fieldset

agentrickard committed:
Adds form facets settings

agentrickard committed:
Finishes backport of D8 options to D7.

agentrickard committed:
Updates copyright

agentrickard committed:
Fixes autocomplete handling for the form

agentrickard commented:
I have most of the settings backported and the app JS/CSS updated to v2.0.

![Screen Shot 2019-03-20 at 4 58 09 PM](https://user-images.githubusercontent.com/360988/54718678-703c6e00-4b31-11e9-916d-1e506c43c170.png)

I have not yet touched the block code.

agentrickard commented:
It's also not clear that I am handling the 'hide' options correctly.

agentrickard committed:
temporary fix for hiding facets.

agentrickard commented on search_api_federated_solr.module:
```diff
@@ -353,31 +353,31 @@ function search_api_federated_solr_admin($form, &$form_state) {
       'id' => ['site-name-property'],
     ],
-    '#default_value' => variable_get('search_api_federated_solr_has_site_name_property') ? 'true' : NULL,
+    '#default_value' => 'true', #variable_get('search_api_federated_solr_has_site_name_property'),
```
On Drupal 7 we don't have an "easy" way of determining what fields are present, like we do in D8, so I defaulted all of these to 'true' for the time being.

In D8, we derive these values from `$type_property = $index_config->get('field_settings.federated_type');`

agentrickard committed:
Fixes handling of default field settings

agentrickard committed:
Adds install/update hook

agentrickard committed:
Removes text reference to non-existent feature.

agentrickard committed:
Changes menu title to Federated Search App

agentrickard committed:
Adds autocomplete settings for proxy

agentrickard committed:
Moves admin form to an include file.

agentrickard committed:
Handles proixy settings

agentrickard committed:
Adds the proxy callback

agentrickard committed:
Test proixy handler

agentrickard committed:
Moves the page callback to an include.

agentrickard committed:
Proxy testing.

agentrickard committed:
Partial handling of proxy.

agentrickard commented on search_api_federated_solr.proxy.inc:
```diff
@@ -0,0 +1,65 @@
+ * Uses the selected index server's backend connector to execute
+ * a select query on the index based on request qs params passed from the app.
+ */
+function search_api_federated_solr_proxy() {
```
This is the part that I just don't quite understand. THe D7 and D8 code for SearchAPISolr is just different enough to be confusing.

agentrickard committed:
Handles block configuration.

agentrickard committed:
Uninstall variables

agentrickard committed:
Incorporates PR #70 to this work.

agentrickard committed:
Fix proxy url

agentrickard committed:
Adds @TODO for block autocomplete.

agentrickard committed:
Starts backport of custom autocomplete js

agentrickard committed:
Additional handling of block autocomplete. Needs work.

agentrickard committed:
Initial backport of block autocomplete jQuery

agentrickard committed:
Autocomplete returns proper results.

agentrickard committed:
Remove debug

agentrickard committed:
Mocu closer to working proxy.

agentrickard committed:
Autocomplete form proxy works

agentrickard committed:
Adds TODO

agentrickard committed:
Test for query defaults.

agentrickard committed:
More testing.

agentrickard committed:
Ensure JSON request from autocomplete block.

agentrickard committed:
Fixes .on calls in js

agentrickard committed:
Gets the block autocomplete to return results. Note @TODO items

agentrickard committed:
Updates React application references.

agentrickard committed:
Check search params.

jesconstantine committed:
Turn off browser autocomplete on block search form

jesconstantine committed:
Fix setting name on check for block autocomplete enabled config value

jesconstantine committed:
Ensure default options overwrite empty block config values

This was preventing results from showing up in the block autocomplete form because suggestionRows was set to and not being overwritten

jesconstantine committed:
Determine the autocomplete endpoint url + params based on block config

jesconstantine committed:
Ensure that proxy url is used when enabled, even if there was direct url

jesconstantine committed:
Add note re: generating menu link vs hard coding

jesconstantine committed:
Ensure qs params do not get appended to autocomplete direct urls

jesconstantine committed:
Fix syntax for sort param (default value only)

jesconstantine committed:
Ensure highlighting data returned in proxy response

jesconstantine committed:
Get, validate and set default query fields for proxy query

jesconstantine committed:
TEMP: root-1002: update the react app in the module

jesconstantine committed:
root-1002: set the flag indicating D7 site

jesconstantine committed:
root-1002: calculate autocomplete proxy url vs hard coding

jesconstantine committed:
platform-1002: add qs parse helper to handle multiple facet.field params

jesconstantine committed:
platform-1002: determine and generate appropriate search app endpoint

jesconstantine committed:
root-1002: make search proxy work with search results page

jesconstantine committed:
platform-1002: comment out code creating SearchApi Query object

jesconstantine committed:
platform-1002: namespace the qs parser function

jesconstantine committed:
platform-1002: make filter queries work with proxy

jesconstantine committed:
platform-1002: make proxy support show results for empty query config

jesconstantine committed:
platform-1002: make sort work with proxy

jesconstantine committed:
platform-1002: urldecode the entire qs at once

jesconstantine committed:
platform-1002: code cleanup

jesconstantine committed:
platform-1002: ensure facet_counts->facet_fields exists in result object

agentrickard commented on search_api_federated_solr.module:
```diff
@@ -757,3 +764,49 @@ function search_api_federated_solr_proxy_params() {
+ * @return string
+ *   URL for the endpoint to be used for query requests.
+ */
+function search_api_federated_solr_get_endpoint_url($proxy_is_disabled, $direct_url, $qs = '') {
```
I had a naiive version of this function in the module file, so it likely needs to be removed (if it hasn't already).

jesconstantine replied:
I think I saw your function which gets the server URL, which is called in this function.

agentrickard commented on search_api_federated_solr.module:
```diff
@@ -785,11 +785,8 @@ function search_api_federated_solr_proxy_params() {
-  ];
   // Default to proxy url.
+  // @TODO is there a way to generate this from the menu?
   $endpoint_url = $base_url . '/search-api-federated-solr/search';
```
No. The other option is to use `url()` and pass `$options['absolute'] = TRUE;` Using `url()` is technically safer.

agentrickard commented on search_api_federated_solr.proxy.inc:
```diff
@@ -30,26 +30,40 @@ function search_api_federated_solr_proxy() {
-  // Get fields from Solr?
-  $cf = $connection->getFields();
+  // Get all full text fields from the index.
+  $full_text_index_fields = $index->getFullTextFields();
```
For some reason, I couldn't find this method in the D7 code...

agentrickard commented on search_api_federated_solr.page.inc:
```diff
@@ -29,6 +29,9 @@ function search_api_federated_solr_config_json() {
   }
+
+  // Set the D7 flag.
+  $federated_search_app_config['isD7'] = TRUE;
```
This var is a new signal to the React app?

jesconstantine replied:
Yep, this is the approach that you, Byron, and I discussed to ensure that the react app sends `search` instead of `q` as the main query param key.

jesconstantine committed:
platform-1002: use drupal url method to compute proxy url

jesconstantine committed:
platform-1002: set the appropriate basic auth credentials for the block

jesconstantine committed:
platform-1002: ensure proxy works with acquia search

jesconstantine committed:
platform-1002: only override search service + connection for acquia search servers

jesconstantine committed:
platform-1002: enable flag for default query fields to not be validated

Adds a field to the search api server edit form under "advanced" and a submission handler which sets a module variable with the input value.

jesconstantine committed:
platform-1002: add drush command to return mapping of index->solr fields

jesconstantine commented on search_api_federated_solr.admin.inc:
```diff
@@ -0,0 +1,521 @@
+  ];
+
+  $form['autocomplete']['direct']['search_api_federated_solr_autocomplete_url'] = [
+    '#type' => 'textfield',
```
Do we need to validate this as a URL (either absolute or relative) at all?

jesconstantine replied:
DOes this suffice? See: Does this suffice? See: https://github.com/palantirnet/search_api_federated_solr/pull/82/commits/eca5607ac99882424c96bde3ef20773422da66f9

jesconstantine commented on search_api_federated_solr.module:
```diff
@@ -115,6 +112,249 @@ function search_api_federated_solr_block_view($delta = '') {
+  ];
+
+  $form['autocomplete']['direct']['search_api_federated_solr_autocomplete_block_url'] = [
+    '#type' => 'textfield',
```
Do we need to validate this as a URL (either absolute or relative) at all?

agentrickard replied:
We likely should validate that it's a URL and reachable.

jesconstantine replied:
Does this suffice? See: https://github.com/palantirnet/search_api_federated_solr/pull/82/commits/eca5607ac99882424c96bde3ef20773422da66f9

jesconstantine committed:
platform-1002: only validate default fields against schema if configured

jesconstantine committed:
platform-1002: validate autocomplete + block autocomplete endpoint url

jesconstantine committed:
platform-1002: update query field param to 'qf' (vs 'df')

jesconstantine committed:
platform-1002: use better variable names for query fields, validation

jesconstantine committed:
platform-1002: enable flag to turn on query debug

jesconstantine committed:
platform-1002: add default query field docs

jesconstantine committed:
platform-1002: update docblock for drush command

jesconstantine committed:
Bump react dependency to official release 2.1.1

jesconstantine committed:
Transpile autocomplete js to es5 for IE

jesconstantine committed:
platform-1002: ensure that default help autocomplete title text is used

jesconstantine committed:
Code cleanup

jesconstantine committed:
Platform-1002: add aria live functionality for results loaded text

jesconstantine committed:
platform-1002: add polyfill for Object.assign

jesconstantine committed:
platform-1002: pass regex pattern directly into string.replace()

Since IE does not support RegExp() constructor

jesconstantine committed:
platform-1002: pull in react app version which works in IE 11

jesconstantine committed:
platform-1002: enable screenreaders to announce header autocomplete options

jesconstantine committed:
platform-1002: add debug docs

jesconstantine committed:
platform-1002: add context to disable query field validation docs

byrond committed:
allow site name to be empty to use system site name and remove default value

agentrickard committed:
Handles URLs properly for D7 views.

agentrickard committed:
Adds 'search_api_federated_solr_set_search_site' to variable list

agentrickard committed:
Properly filter the proxy when site name is restricted.

jesconstantine commented on search_api_federated_solr.proxy.inc:
```diff
@@ -0,0 +1,172 @@
+    $query_fields = array_values($query_fields_map);
+  }
+
+  // If site search is restricted, enforce it here.
```
I think we can only enforce this if site search is restricted *and hidden*.  Otherwise, it's possible (as in UMich's case) that this could be a subsequent search and the site name facet has been unchecked.  See the description for the default site name field: https://github.com/palantirnet/search_api_federated_solr/pull/82/files#diff-819816c1f9a4704c4a9e9f3e53e2d832R185

agentrickard replied:
that's fine.

jesconstantine commented on search_api_federated_solr.proxy.inc:
```diff
@@ -0,0 +1,172 @@
+    }
+    foreach ($params['fq'] as $index => $element) {
+      if (substr_count($element, 'sm_site_name:') > 0) {
+        unset($params['fq'][$index]);
```
Can you walk me through this... is this unsetting the site name if it already exists as a qs param?

agentrickard replied:
Yes, to ensure we aren't sending more than one value to the server.

jesconstantine commented on search_api_federated_solr.proxy.inc:
```diff
@@ -0,0 +1,172 @@
+      }
+    }
+    // @TODO: This value might be configurable? Get from SearchAPI config.
+    $params['fq'][] = 'sm_site_name:("' . variable_get('site_name') . '")';
```
Yes, this should just equal the value of the site name property on the index.

agentrickard replied:
I just fixed that part in a subsequent commit.

agentrickard committed:
Handles the site name field properly in the proxy.

agentrickard committed:
Only restrict site if the site facet is hidden.

agentrickard commented on search_api_federated_solr.proxy.inc:
```diff
@@ -73,7 +71,8 @@ dpm($index);
   // If site search is restricted, enforce it here.
-  $restrict_site = variable_get('search_api_federated_solr_set_search_site', 0);
+  $restrict_site = variable_get('search_api_federated_solr_set_search_site', 0)
+                && variable_get('search_api_federated_solr_hide_site_name', 0);
```
@jesconstantine  Since this _only_ works with the proxy right now, should we force the proxy to be used in configuration, or just set a note in the UI that the proxy must be enabled?

jesconstantine replied:
The only time it won't work is if the site is not using the proxy AND the search page is browsed to directly, right?  I think we can make a note in the UI with that restriction.

I also envision a near future where we only support using the proxy (or a view for autocomplete).

agentrickard committed:
Fixes issue with fq index. Adds menu description.

agentrickard committed:
Adds help text to configuration of site filter.

agentrickard committed:
Fixes variable name collision.

agentrickard committed:
Push comment.
